profiling/src/collections/string_storage.rs

```diff
@@ -0,0 +1,148 @@
+    cached_seq_num: Cell<Option<StringId>>,
+    usage_count: Cell<u32>,
+}
+
+pub struct ManagedStringStorage {
+    next_id: u32,
+    id_to_data: HashMap<u32, ManagedStringData, BuildHasherDefault<rustc_hash::FxHasher>>,
+    str_to_id: HashMap<Rc<str>, u32, BuildHasherDefault<rustc_hash::FxHasher>>,
+    current_gen: u32,
+}
+
+impl ManagedStringStorage {
+    pub fn new() -> Self {
+        let mut storage = ManagedStringStorage {
+            next_id: 0,
+            id_to_data: Default::default(),
+            str_to_id: Default::default(),
+            current_gen: 0,
+        };
+        // Ensure empty string gets id 0 and always has usage > 0 so it's always retained
+        storage.intern("");
+        storage
+    }
+
+    pub fn advance_gen(&mut self) {
+        self.id_to_data.retain(|_, data| {
+            /*let used_in_this_gen = data
+                .last_usage_gen
+                .get()
+                .map_or(false, |last_usage_gen| last_usage_gen == self.current_gen);
+            if used_in_this_gen || *id == 0 {
+                // Empty string (id = 0) or anything that was used in the gen
+                // we are now closing, is kept alive
+                return true;
+            }*/
+            if data.usage_count.get() > 0 {
+                return true;
+            }
+
+            self.str_to_id.remove_entry(&data.str);
+            false
+        });
+        self.current_gen += 1;
+    }
+
+    pub fn intern(&mut self, item: &str) -> u32 {
+        let entry = self.str_to_id.get_key_value(item);
+        match entry {
+            Some((_, id)) => {
+                let usage_count = &self
+                    .id_to_data
+                    .get(id)
+                    .expect("id_to_data and str_to_id should be in sync")
+                    .usage_count;
+                usage_count.set(usage_count.get() + 1);
+                *id
+            }
+            None => {
+                let id = self.next_id;
+                let str: Rc<str> = item.into();
+                let data = ManagedStringData {
+                    str: str.clone(),
+                    last_usage_gen: Cell::new(None),
+                    cached_seq_num_for: Cell::new(None),
+                    cached_seq_num: Cell::new(None),
+                    usage_count: Cell::new(1),
+                };
+                self.next_id = self.next_id.checked_add(1).expect("Ran out of string ids!");
+                let old_value = self.str_to_id.insert(str.clone(), id);
+                debug_assert_eq!(old_value, None);
+                let old_value = self.id_to_data.insert(id, data);
+                debug_assert_eq!(old_value, None);
+                id
+            }
+        }
+    }
+
+    pub fn unintern(&self, id: u32) {
+        if id == 0 {
+            panic!("For performance reasons, unintern should not be called with id == 0. Please hardcode a fast path check in the caller to avoid this call")
+        }
```

Comment:
This can be removed right?

Reply:
Actually, no!

This check is not about the empty string existing or not, it's about pushing callers to do the `if id == 0 return` checks as close to the client code as possible, rather than "eating" all the overhead of coming all the way up here (including all of the lock acquisition that it may entail) just to go "oh, it's a zero, nothing to do!".

So it's the equivalent of putting a big banner on the lobby saying "don't bother going up the stairs all the way to the top if you have an empty string, save yourself some work".